src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java

```diff
@@ -0,0 +1,132 @@
+     * @param dictionary dictionary for position comparison.
+     */
+    public ShardingIterator(final Iterator<T> iterator,
+            final List<ShardBoundary> boundaries,
+            final SAMSequenceDictionary dictionary) {
+        // check arguments
+        Utils.validateArg(Utils.nonNull(iterator, "null iterator").hasNext(), "empty iterator");
+        Utils.nonEmpty(boundaries, "Empty shard boundaries");
+        Utils.nonNull(dictionary, "null dictionary");
+
+        // TODO: avoid new object creation for already PeekableIterator and LinkedList args
+        // store the iterator and allow peek()
+        this.it = new PeekableIterator<>(iterator);
+        // TODO: avoid new object creation for already LinkedList and LinkedList args
+        // convert the boundaries into a queue
+        this.shards = new LinkedList<>(boundaries);
+        // store the dictionary
+        this.dictionary = dictionary;
+        // create a new queue for the next data
+        this.nextData = new LinkedList<>();
+
+        // start the iteration
+        advance();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return !shards.isEmpty();
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Note: for overlapping shards, the objects stored in two (or more) consecutive shards are
+     * references and not clones. If modified in the previous shard, the state would be kept in the
+     * next returned value. It is up to the caller to make copies if necessary to avoid state in the
+     * next iteration.
+     */
+    @Override
+    public Shard<T> next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+        // copying the deque to being able to remove unnecesary data and fill in with the next
+        final Shard<T> next = new ShardBoundaryShard<T>(shards.poll(),
+                nextData.isEmpty() ? Collections.emptyList() : new LinkedList<>(nextData));
+        advance();
+        return next;
+    }
+
+    // advance to the next data
+    private void advance() {
+        // if we already finished the iteration, we clean up the data
+        if (shards.isEmpty()) {
+            // TODO: should we also close the iterator?
+            nextData.clear();
+        } else {
+            // first we empty the queue if it has non-overlapping data with the next shard
+            removeNonOverlapping();
+            // then, fill in the data
+            fillQueueWithOverlapping();
+        }
+    }
+
+    // assumes that shards are not empty
+    private void removeNonOverlapping() {
+        while (!nextData.isEmpty() && !nextData.peek()
+                .overlaps(shards.peek().getPaddedInterval())) {
+            nextData.poll();
+        }
+    }
+
+    // assumes that shards are not empty
+    private void fillQueueWithOverlapping() {
+        while (it.hasNext()) {
+            // if this one overlaps
+            if (it.peek().overlaps(shards.peek().getPaddedInterval())) {
```

Comment:
Your iterator will not work with unmapped reads, which implies that you should have the `MappedReadFilter` on by default in your walker base class (in addition to the `WellformedReadFilter`, which is already on).

Reply:
I added already the `MappedReadFilter`. I might consider to add unmapped reads too, but not at the moment.